drivers/flash/flash_npcx_fiu_nor.c

```diff
@@ -0,0 +1,550 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_fiu_nor
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/flash/npcx_flash_api_ex.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/flash_controller/npcx_fiu_qspi.h>
+#include <soc.h>
+
+#include "flash_npcx_fiu_qspi.h"
+#include "spi_nor.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(flash_npcx_fiu_nor, CONFIG_FLASH_LOG_LEVEL);
+
+/* Device config */
+struct flash_npcx_nor_config {
+	/* QSPI bus device for mutex control and bus configuration */
+	const struct device *qspi_bus;
+	/* Mapped address for flash read via direct access */
+	uintptr_t mapped_addr;
+	/* Size of nor device in bytes, from size property */
+	uint32_t flash_size;
+	/* Minimum size for flash erase */
+	uint32_t min_erase_size;
+	/* Maximum chip erase time-out in ms */
+	uint32_t max_timeout;
+	/* SPI Nor device configuration on QSPI bus */
+	struct npcx_qspi_cfg qspi_cfg;
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	struct flash_pages_layout layout;
+#endif
+};
+
+/* Device data */
+struct flash_npcx_nor_data {
+	/* Specific control operation for Quad-SPI Nor Flash */
+	uint32_t operation;
+};
+
+static const struct flash_parameters flash_npcx_parameters = {
+	.write_block_size = 1,
+	.erase_value = 0xff,
+};
+
+#define DT_INST_QUAD_EN_PROP_OR(inst)						\
+	COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, quad_enable_requirements),	\
+		    (_CONCAT(JESD216_DW15_QER_VAL_,				\
+		     DT_INST_STRING_TOKEN(inst, quad_enable_requirements))),	\
+		    ((JESD216_DW15_QER_VAL_NONE)))
+
+static int flash_npcx_uma_transceive(const struct device *dev, struct npcx_uma_cfg *cfg,
+				     uint32_t flags)
+{
+	const struct flash_npcx_nor_config *config = dev->config;
+	uint32_t operation = ((struct flash_npcx_nor_data *)dev->data)->operation;
```

Comment:
Maybe use an intermediate variable rather than the inline casting?

Reply:
Done. I also declare `operation` parameter as `const` to make sure it won't change in `qspi_npcx_fiu_mutex_lock_configure()`

Reply:
Done.

Reply:
Sorry I meant: use an intermediate variable for `dev->data`, like

```c
struct flash_npcx_nor_data *data = dev->data;
...
oper = data->operation
```

Reply:
Oops. Sorry for the misunderstanding. Since there's a variable naming conflict with `data` in `read` flash API function, I replace `data` with `dev_data` in the newest PR. Any further comment is appreciated.